sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/ParameterHandler.scala

```diff
@@ -0,0 +1,240 @@
+ * limitations under the License.
+ */
+package org.apache.spark.sql.catalyst.parser
+
+import org.apache.spark.sql.catalyst.expressions.{Expression, Literal}
+import org.apache.spark.sql.catalyst.parser.SubstituteParamsParser
+import org.apache.spark.sql.catalyst.trees.CurrentOrigin
+import org.apache.spark.sql.catalyst.util.LiteralToSqlConverter
+
+/**
+ * Handler for parameter substitution across different Spark SQL contexts.
+ *
+ * This object consolidates the common parameter handling logic used by SparkSqlParser,
+ * SparkConnectPlanner, and ExecuteImmediate. It provides a single, consistent API
+ * for all parameter substitution operations in Spark SQL.
+ *
+ * Key features:
+ * - Automatic parameter type detection (named vs positional)
+ * - Uses CompoundOrSingleStatement parsing for all SQL constructs
+ * - Consistent error handling and validation
+ * - Support for complex data types (arrays, maps, nested structures)
+ * - Thread-safe operations with position-aware error context
+ *
+ * The handler integrates with the parser through callback mechanisms stored in
+ * CurrentOrigin to ensure error positions are correctly mapped back to the original SQL text.
+ *
+ * @example Basic usage:
+ * {{{
+ * val context = NamedParameterContext(Map("param1" -> Literal(42)))
+ * val result = ParameterHandler.substituteParameters("SELECT :param1", context)
+ * // result: "SELECT 42"
+ * }}}
+ *
+ * @see [[SubstituteParamsParser]] for the underlying parameter substitution logic
+ */
+object ParameterHandler {
+
+  // Compiled regex pattern for efficient parameter marker detection.
+  private val parameterMarkerPattern = java.util.regex.Pattern.compile("[?:]")
+
+  /**
+   * Performs parameter substitution and stores the position mapper in CurrentOrigin.
+   * The position mapper enables accurate error reporting by mapping positions in the
+   * substituted SQL back to the original SQL text.
+   *
+   * @param sqlText The SQL text containing parameter markers
+   * @param namedParams Optional named parameters map
+   * @param positionalParams Optional positional parameters list
+   * @return The SQL text with parameters substituted
+   */
+  private def performSubstitution(
+      sqlText: String,
+      namedParams: Map[String, String] = Map.empty,
+      positionalParams: List[String] = List.empty): String = {
+
+    // Optimize: Skip parsing if no parameter markers are present in the SQL text.
+    if (!parameterMarkerPattern.matcher(sqlText).find()) {
+      return sqlText
+    }
+
+    val substitutor = new SubstituteParamsParser()
+    val (substituted, _, positionMapper) = substitutor.substitute(sqlText,
+      namedParams = namedParams, positionalParams = positionalParams)
+
+    // Store the position mapper to enable accurate error position reporting in the original SQL.
+    val currentOrigin = CurrentOrigin.get
+    CurrentOrigin.set(currentOrigin.copy(positionMapper = Some(positionMapper)))
```

Comment:
This is a side effect that updates a global variable. I'm a bit worried about it as we never restore it.

Reply:
I think it's safer to return the position mapper to the caller, and let caller to set up the `CurrentOrigin` before invoking the main parser, and restore it afterwards.